memory.go

```diff
@@ -0,0 +1,280 @@
+// is guaranteed to access bpf map memory. When all references are gone, the
+// allocation's finalizer will run and unmap the bpf map's memory.
+
+//go:linkname heapObjectsCanMove runtime.heapObjectsCanMove
+func heapObjectsCanMove() bool
+
+var ErrReadOnly = errors.New("resource is read-only")
+
+// Memory implements accessing a Map's memory without making any syscalls.
+// Pay attention to the difference between Go and C struct alignment rules. Use
+// [structs.HostLayout] on supported Go versions to help with alignment.
+//
+// Note on memory coherence: avoid using packed structs in memory shared between
+// user space and eBPF C programs. This drops a struct's memory alignment to 1,
+// forcing the compiler to use single-byte loads and stores for field accesses.
+// This may lead to partially-written data to be observed from user space.
+//
+// On most architectures, the memmove implementation used by Go's copy() will
+// access data in word-sized chunks. If paired with a matching access pattern on
+// the eBPF C side (and if using default memory alignment), accessing shared
+// memory without atomics or other synchronization primitives should be sound
+// for individual values. For accesses beyond a single value, the usual
+// concurrent programming rules apply.
+type Memory struct {
+	b  []byte
+	ro bool
+}
+
+func newMemory(fd, size int, ro bool) (*Memory, error) {
+	// Some architectures need the size to be page-aligned.
+	if internal.Align(size, os.Getpagesize()) != size {
+		return nil, fmt.Errorf("memory: must be a multiple of page size (requested %d bytes)", size)
+	}
+
+	// Allocate a page-aligned span of memory on the Go heap.
+	alloc, err := allocate(size)
+	if err != nil {
+		return nil, fmt.Errorf("allocating memory: %w", err)
+	}
+
+	flags := unix.PROT_READ | unix.PROT_WRITE
+	if ro {
+		flags = unix.PROT_READ
+	}
+
+	// Map the bpf map memory over a page-aligned allocation on the Go heap.
+	if err := mapmap(fd, alloc, size, flags); err != nil {
+		return nil, fmt.Errorf("mapping memory: %w", err)
+	}
+
+	mm := &Memory{
+		unsafe.Slice((*byte)(alloc), size),
+		ro,
+	}
+
+	return mm, nil
+}
+
+// allocate returns a pointer to a page-aligned section of memory on the Go
+// heap, managed by the runtime.
+func allocate(size int) (unsafe.Pointer, error) {
+	// Memory-mapping over a piece of the Go heap is unsafe when the GC can
+	// randomly decide to move objects around, in which case the mapped region
+	// will not move along with it.
+	if heapObjectsCanMove() {
+		return nil, errors.New("this Go runtime has a moving garbage collector")
+	}
+
+	if size == 0 {
+		return nil, errors.New("size must be greater than 0")
+	}
+
+	// Request at least two pages of memory from the runtime to ensure we can
+	// align the requested allocation to a page boundary. This is needed for
+	// MAP_FIXED and makes sure we don't mmap over some other allocation on the Go
+	// heap.
+	size = internal.Align(size+os.Getpagesize(), os.Getpagesize())
```

Comment:
No, this wastes only one page. If the bpf map is one page, we request two pages from the runtime. On a 16k page architecture, this would 'waste' 16k, not 32k. The other 16k would be mapped to kernel memory. If the slice's allocation is page-aligned (which it should be under the current Go allocator), the second page will never be touched and will technically never get dirtied, only 'consuming' virtual memory.